dbms/src/Interpreters/Aggregator.cpp

```diff
@@ -678,43 +716,65 @@ ALWAYS_INLINE void Aggregator::executeImplBatch(
 
-        auto emplace_result = state.emplaceKey(method.data, i, *aggregates_pool, sort_key_containers);
+            allow_exception = true;
 
-        /// If a new key is inserted, initialize the states of the aggregate functions, and possibly something related to the key.
-        if (emplace_result.isInserted())
-        {
-            /// exception-safety - if you can not allocate memory or create states, then destructors will not be called.
-            emplace_result.setMapped(nullptr);
+            auto emplace_result = state.emplaceKey(method.data, i, *aggregates_pool, sort_key_containers);
 
-            aggregate_data = aggregates_pool->alignedAlloc(total_size_of_aggregate_states, align_aggregate_states);
-            createAggregateStates(aggregate_data);
+            allow_exception = false;
 
-            emplace_result.setMapped(aggregate_data);
-        }
-        else
-            aggregate_data = emplace_result.getMapped();
+            /// If a new key is inserted, initialize the states of the aggregate functions, and possibly something related to the key.
+            if (emplace_result.isInserted())
+            {
+                /// exception-safety - if you can not allocate memory or create states, then destructors will not be called.
+                emplace_result.setMapped(nullptr);
+
+                aggregate_data = aggregates_pool->alignedAlloc(total_size_of_aggregate_states, align_aggregate_states);
+                createAggregateStates(aggregate_data);
+
+                emplace_result.setMapped(aggregate_data);
+            }
+            else
+                aggregate_data = emplace_result.getMapped();
 
-        places[i - agg_process_info.start_row] = aggregate_data;
+            places[i - agg_process_info.start_row] = aggregate_data;
+            processed_rows = i;
+        }
+    }
+    catch (ResizeException &)
+    {
+        LOG_INFO(
+            log,
+            "HashTable resize throw ResizeException since the data is already marked for spill, allow_exception: {}",
+            allow_exception);
+        if unlikely (!allow_exception)
+            throw;
     }
 
-    /// Add values to the aggregate functions.
-    for (AggregateFunctionInstruction * inst = agg_process_info.aggregate_functions_instructions.data(); inst->that;
-         ++inst)
+    if (processed_rows != std::numeric_limits<size_t>::max())
     {
-        inst->batch_that->addBatch(
-            agg_process_info.start_row,
-            agg_size,
-            places.get(),
-            inst->state_offset,
-            inst->batch_arguments,
-            aggregates_pool);
+        /// Add values to the aggregate functions.
+        for (AggregateFunctionInstruction * inst = agg_process_info.aggregate_functions_instructions.data(); inst->that;
+             ++inst)
+        {
+            inst->batch_that->addBatch(
+                agg_process_info.start_row,
+                processed_rows - agg_process_info.start_row + 1,
```

Comment:
```suggestion
                agg_size,
```
If `processed_rows != std::numeric_limits<size_t>::max()`, it seems that `processed_rows` must be `agg_process_info.start_row + 1 + agg_size`

Reply:
Must use `processed_rows - agg_process_info.start_row + 1` here, because the basic idea is:
- `processed_rows` is used to record how many rows are already insert its key to the hash table
- during insert, it may throw ResizeException, if exception happens, the processed_rows is pointing the last row that have their key in the hash table
- if `allow_exception` is true when `ResizeException` happens, the aggregator will aggregate data in [start_rows, processed_rows], then update `start_rows` to `processed_rows + 1`, so next time in `executeOnBlock`, it will not re-aggregate the ones that is already aggregated.
- `allow_exception` is used to make sure the valid exception only comes when insert key to hash table, for example, allocating memory in pool does not allow exception.